stochatreat/stochatreat.py

```diff
@@ -150,75 +160,73 @@ def stochatreat(data: pd.DataFrame,
-            # assign random values
-            slize['rand'] = R.uniform(size=len(slize))
-            # sort by random
-            slize = slize.sort_values(by='rand')
-            # drop the rand column
-            slize = slize.drop(columns='rand')
-            # reset index in order to keep original id
-            slize = slize.reset_index(drop=True)
-            # assign treatment by index
-            for i, treat in enumerate(ts):
-                if treat == 0:
-                    slize.loc[:locs[i], 'treat'] = treat
-                else:
-                    slize.loc[locs[i - 1]:locs[i], 'treat'] = treat
-            new_slize = slize.copy()
-
-        # if there are any misfits
-        elif n_misfit > 0:
-            # separate groups between adherents and misfits
-            adherents = slize.iloc[:n_belong].copy()
-            misfits = slize.iloc[n_belong:].copy()
-
-            # assign adherents
-            adherents['rand'] = R.uniform(size=len(adherents))
-            # sort by random
-            adherents = adherents.sort_values(by='rand')
-            # drop the rand column
-            adherents = adherents.drop(columns='rand')
-            # reset index in order to keep original id
-            adherents = adherents.reset_index(drop=True)
-            # assign treatment by index
-            for i, treat in enumerate(ts):
-                if treat == 0:
-                    adherents.loc[:locs[i], 'treat'] = treat
-                else:
-                    adherents.loc[locs[i - 1]:locs[i], 'treat'] = treat
-            new_slize.append(adherents)
-
-            # assign misfits
-            misfits['treat'] = R.choice(range(treats),
-                                        size=n_misfit,
-                                        p=probs)
-            new_slize.append(misfits)
-            new_slize = pd.concat(new_slize)
+        # assign treatments in wanted proportions
+        n_repeat_mask = block_size // lcm_prob_denominators
+        block_treatments = np.repeat(treat_mask, n_repeat_mask)
+
+        # deal with misfits
+        n_misfit = block_size % lcm_prob_denominators
+
+        if n_misfit > 0:
+            if misfit_strategy == "stratum":
+                misfit_treatments = R.choice(
+                    range(treats),
+                    size=n_misfit,
+                    p=probs
+                )
+                block_treatments = np.r_[block_treatments, misfit_treatments]
+            elif misfit_strategy == "global":
+                misfit_data = slize.sample(n_misfit)
+                slize = slize.drop(index=misfit_data.index)
+                global_misfits.append(misfit_data)
+            else:
+                raise ValueError("strategy must be one of 'stratum' or 'global'")
+
+        np.random.shuffle(block_treatments)
+        slize['treat'] = block_treatments
 
         # append blocks together
-        slizes.append(new_slize)
+        slizes.append(slize)
```

Comment:
Is the best way to avoid the code duplication to (optionally) create the misfit block first, then run through all blocks and do the two rounds of assignments?

Reply:
As I was writing an alternative I found your solution better, so yes probably a good way to avoid duplication!